File: 1603.design-parking-system.cpp
/*
 * @lc app=leetcode id=1603 lang=cpp
 *
 * [1603] Design Parking System
 */

// @lc code=start
class ParkingSystem {
 public:
  ParkingSystem(int big, int medium, int small)
      : big_(big), medium_(medium), small_(small) {}

  bool addCar(int carType) {
    switch (carType) {
      case 1:
        if (big_ > 0) {
          big_--;
          return true;
        }
        break;
      case 2:
        if (medium_ > 0) {
          medium_--;
          return true;
        }
        break;
      case 3:
        if (small_ > 0) {
          small_--;
          return true;
        }
        break;
    }
    return false;
  }

 private:
  int big_;
  int medium_;
  int small_;
};

/**
 * Your ParkingSystem object will be instantiated and called as such:
 * ParkingSystem* obj = new ParkingSystem(big, medium, small);
 * bool param_1 = obj->addCar(carType);
 */
// @lc code=end

File: 1603.design-parking-system.rs
/*
 * @lc app=leetcode id=1603 lang=rust
 *
 * [1603] Design Parking System
 */

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
struct ParkingSystem {
    v: Vec<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl ParkingSystem {
    fn new(big: i32, medium: i32, small: i32) -> Self {
        ParkingSystem {
            v: vec![0, big, medium, small],
        }
    }

    fn add_car(&mut self, car_type: i32) -> bool {
        if self.v[car_type as usize] == 0 {
            return false;
        }
        self.v[car_type as usize] -= 1;
        true
    }
}

/**
 * Your ParkingSystem object will be instantiated and called as such:
 * let obj = ParkingSystem::new(big, medium, small);
 * let ret_1: bool = obj.add_car(carType);
 */
// @lc code=end

#[test]
fn test1() {
    let mut ps = ParkingSystem::new(1, 1, 0);
    assert!(ps.add_car(1));
    assert!(ps.add_car(2));
    assert!(!ps.add_car(3));
    assert!(!ps.add_car(1));
}
